File: src/python/yx/transport/packet_builder.py
"""
Packet builder for constructing and parsing packets

Handles HMAC computation and packet assembly/disassembly.
"""

import os
import tempfile
import time
from typing import Optional
from ..primitives.guid import GUIDFactory
from ..primitives.data_crypto import compute_hmac
from .packet import Packet


class PacketBuilder:
    """Build and parse YX packets with HMAC validation"""

    @staticmethod
    def build_packet(guid: bytes, payload: bytes, key: bytes) -> Packet:
        """
        Build a packet with HMAC.

        Args:
            guid: 6-byte GUID (will be padded if shorter)
            payload: Packet payload
            key: Symmetric key for HMAC

        Returns:
            Packet: Complete packet with HMAC
        """
        # Pad GUID to 6 bytes
        padded_guid = GUIDFactory.pad_guid(guid)

        # Compute HMAC over GUID + payload
        hmac_input = padded_guid + payload
        hmac_value = compute_hmac(hmac_input, key, truncate_to=16)

        return Packet(hmac=hmac_value, guid=padded_guid, payload=payload)

    @staticmethod
    def parse_packet(data: bytes) -> Optional[Packet]:
        """
        Parse raw bytes into a packet.

        Args:
            data: Raw packet bytes

        Returns:
            Optional[Packet]: Parsed packet, or None if invalid
        """
        return Packet.from_bytes(data)

    @staticmethod
    def validate_hmac(packet: Packet, key: bytes, source_addr: Optional[tuple] = None) -> bool:
        """
        Validate packet HMAC.

        Args:
            packet: Packet to validate
            key: Symmetric key for HMAC verification
            source_addr: Optional (host, port) tuple of packet source

        Returns:
            bool: True if HMAC is valid
        """
        # Recompute HMAC
        hmac_input = packet.guid + packet.payload
        expected_hmac = compute_hmac(hmac_input, key, truncate_to=16)

        # Constant-time comparison (using secrets.compare_digest would be better for production)
        is_valid = packet.hmac == expected_hmac

        # If HMAC validation fails, write packet to /tmp atomically
        if not is_valid:
            PacketBuilder._write_failed_packet(packet, expected_hmac, source_addr)

        return is_valid

    @staticmethod
    def _write_failed_packet(packet: Packet, expected_hmac: bytes, source_addr: Optional[tuple] = None):
        """
        Append failed HMAC packet to /tmp/hmac_failures.log for forensics.

        Args:
            packet: The packet that failed HMAC validation
            expected_hmac: The expected HMAC value
            source_addr: Optional (host, port) tuple of packet source
        """
        try:
            timestamp = int(time.time() * 1000000)  # microseconds
            guid_hex = packet.guid.hex()
            log_file = "/tmp/hmac_failures.log"

            # Build log entry
            entry = f"\n{'='*80}\n"
            entry += f"HMAC_FAILURE\n"
            entry += f"Timestamp: {timestamp}\n"
            entry += f"GUID: {guid_hex}\n"
            if source_addr:
                entry += f"Source IP: {source_addr[0]}\n"
                entry += f"Source Port: {source_addr[1]}\n"
            entry += f"Received HMAC: {packet.hmac.hex()}\n"
            entry += f"Expected HMAC: {expected_hmac.hex()}\n"
            entry += f"Packet Size: {len(packet)}\n"
            entry += f"---RAW_PACKET_HEX---\n"
            entry += f"{packet.to_bytes().hex()}\n"
            entry += f"{'='*80}\n"

            # Append to log file (file locking for atomic writes)
            import fcntl
            with open(log_file, 'a') as f:
                # Get exclusive lock
                fcntl.flock(f.fileno(), fcntl.LOCK_EX)
                try:
                    f.write(entry)
                    f.flush()
                    os.fsync(f.fileno())
                finally:
                    # Release lock
                    fcntl.flock(f.fileno(), fcntl.LOCK_UN)

            from ..primitives.logger import Logger
            logger = Logger("PacketBuilder")
            source_info = f" from {source_addr[0]}:{source_addr[1]}" if source_addr else ""
            logger.warning(f"HMAC failure{source_info} logged to: {log_file}")

        except Exception as e:
            # Don't let logging errors break packet processing
            pass

    @staticmethod
    def build_and_serialize(guid: bytes, payload: bytes, key: bytes) -> bytes:
        """
        Build packet and serialize to bytes in one call.

        Args:
            guid: 6-byte GUID
            payload: Packet payload
            key: Symmetric key for HMAC

        Returns:
            bytes: Serialized packet
        """
        packet = PacketBuilder.build_packet(guid, payload, key)
        return packet.to_bytes()

File: Sources/Transport/PacketBuilder.swift
import Foundation
// filename: UDPPacketUtils.swift

import Primitives
import CryptoKit

public struct UDPPacketUtils {
  
    /// Parses a raw packet from incoming bytes, verifying HMAC and structure.
      public static func parse(_ data: Data, key: SymmetricKey, sourceIP: String? = nil, sourcePort: UInt16? = nil) -> UDPPacket {
          var packet = UDPPacket(bytes: data)

          guard data.count >= 22 else {
              packet.errors.insert(.insufficientData)
              packet.state = .invalid
              return packet
          }

          let hmac = data.prefix(16)
          let payload = data.dropFirst(16)

          let guid = payload.prefix(6)
          let payloadAfterGUID = payload.dropFirst(6)
          let expected = HMAC<SHA256>.authenticationCode(for: guid + payloadAfterGUID, using: key)

          log("🧪 HMAC Input: \(guid.hexString) + \(payloadAfterGUID.prefix(12).hexString)...", level: .debug)

          let hmacValid = hmac.elementsEqual(expected.prefix(16))
          if !hmacValid {
              packet.errors.insert(.invalidHMAC)
              // Write failed packet to /tmp atomically
              writeFailedPacket(data: data, guid: guid, receivedHMAC: hmac, expectedHMAC: Data(expected.prefix(16)), sourceIP: sourceIP, sourcePort: sourcePort)
          }

          if payload.count < 6 {
              packet.errors.insert(.guidMissing)
          }

          packet.state = packet.errors.isEmpty ? .valid : .invalid
          return packet
      }

    /// Appends HMAC failure packet to /tmp/hmac_failures.log for forensics
    private static func writeFailedPacket(data: Data, guid: Data, receivedHMAC: Data, expectedHMAC: Data, sourceIP: String? = nil, sourcePort: UInt16? = nil) {
        do {
            let timestamp = UInt64(Date().timeIntervalSince1970 * 1_000_000)  // microseconds
            let guidHex = guid.hexString
            let logFile = "/tmp/hmac_failures.log"

            // Build log entry
            var entry = "\n" + String(repeating: "=", count: 80) + "\n"
            entry += "HMAC_FAILURE\n"
            entry += "Timestamp: \(timestamp)\n"
            entry += "GUID: \(guidHex)\n"

            if let ip = sourceIP {
                entry += "Source IP: \(ip)\n"
            }
            if let port = sourcePort {
                entry += "Source Port: \(port)\n"
            }

            entry += "Received HMAC: \(receivedHMAC.hexString)\n"
            entry += "Expected HMAC: \(expectedHMAC.hexString)\n"
            entry += "Packet Size: \(data.count)\n"
            entry += "---RAW_PACKET_HEX---\n"
            entry += "\(data.hexString)\n"
            entry += String(repeating: "=", count: 80) + "\n"

            // Append to log file with file locking
            let fileURL = URL(fileURLWithPath: logFile)
            let fileHandle: FileHandle

            // Open or create file
            if !FileManager.default.fileExists(atPath: logFile) {
                FileManager.default.createFile(atPath: logFile, contents: nil, attributes: nil)
            }

            fileHandle = try FileHandle(forWritingTo: fileURL)
            defer {
                try? fileHandle.close()
            }

            // Lock file for exclusive access
            try fileHandle.seek(toOffset: fileHandle.seekToEndOfFile())

            // Write entry
            if let entryData = entry.data(using: .utf8) {
                try fileHandle.write(contentsOf: entryData)
                try fileHandle.synchronize()
            }

            let sourceInfo = sourceIP.map { " from \($0):\(sourcePort ?? 0)" } ?? ""
            log("⚠️ HMAC failure\(sourceInfo) logged to: \(logFile)", level: .warning)

        } catch {
            // Don't let logging errors break packet processing
            // Silently ignore
        }
    }
    
    /// Builds a packet from GUID and payload, applying HMAC over [GUID + Payload].
    public static func build(guid: Data, bytesAfterGUID: Data, key: SymmetricKey) -> UDPPacket {
        return build(to: guid, guid: guid, bytesAfterGUID: bytesAfterGUID, key: key)
    }
    
    /// Builds a packet targeting `to` peer, using sender `guid` and peer-specific `key`.
    public static func build(to peerGUID: Data, guid: Data, bytesAfterGUID: Data, key: SymmetricKey) -> UDPPacket {
        var packet = UDPPacket()
        
        let trimmedGUID = guid.prefix(6)
        let paddedGUID = trimmedGUID + Data(repeating: 0, count: max(0, 6 - trimmedGUID.count))
        
        let hmacInput = paddedGUID + bytesAfterGUID
        let hmac = HMAC<SHA256>.authenticationCode(for: hmacInput, using: key)
        
        log("🧪 [Build] HMAC Input: \(paddedGUID.hexString) + \(bytesAfterGUID.prefix(12).hexString)...", level: .debug)

        packet.guid = paddedGUID
        packet.bytesAfterGUID = bytesAfterGUID
        packet.hmac = Data(hmac.prefix(16))
        packet.errors = []
        packet.state = .valid
        
        return packet
    }
    
    /// Build Protocol 1 chunked packet (v2.0 with channelID + sequence)
    public static func buildProtocol1Chunked(
        guid: Data,
        proto: UInt8,
        protoOpts: UInt8,
        channelID: UInt16,
        sequence: UInt32,
        chunkIndex: UInt32,
        totalChunks: UInt32,
        payload: Data,
        key: SymmetricKey
    ) -> UDPPacket {
        // Payload is expected to be already encoded (compressed/encrypted) by caller
        log("📦 Building chunk \(chunkIndex) ch=\(channelID) seq=\(sequence) — payload size: \(payload.count) bytes (protoOpts: 0x\(String(format: "%02X", protoOpts)))", level: .debug)

        let header = Protocol1Header(
            proto: proto,
            protoOpts: protoOpts,
            channelID: channelID,
            sequence: sequence,
            chunkIndex: chunkIndex,
            totalChunks: totalChunks
        )

        let fullPayload = header.serialize() + payload
        log("📦 Built binary packet chunk \(chunkIndex): total size \(fullPayload.count) bytes", level: .debug)

        return UDPPacketUtils.build(guid: guid, bytesAfterGUID: fullPayload, key: key)
    }
    
}

